src/QsCompiler/SyntaxProcessor/TreeVerification.fs

```diff
@@ -22,10 +22,10 @@ open Microsoft.Quantum.QsCompiler.SyntaxTree
 /// Such diagnostics include in particular warnings for all statement that will never be executed, and errors for misplaced returns statements
 /// Throws an ArgumentException if the statements contain no location information. 
 let AllPathsReturnValueOrFail body = 
-    let diagnostics = new List<Position * QsCompilerDiagnostic>()
+    let diagnostics = ResizeArray<_> ()
     let addDiagnostic diag (stm : QsStatement) = stm.Location |> function
         | Null -> ArgumentException "no location set for the given statement" |> raise
-        | Value loc -> (loc.Offset, loc.Range |> diag) |> diagnostics.Add
+        | Value loc -> loc.Offset + loc.Range |> diag |> diagnostics.Add
```

Comment:
The pattern offset+range seem to repeat quite a bit. What happens if offset isn't added to range, will it still compile? If it will, is there any way to prevent new code from forgetting to add offset?

Reply:
If you forget to add the offset (so you write `loc.Range` instead of `loc.Offset + loc.Range`), it will still compile because the type of both expressions is `Range`, but the range will be incorrect (i.e. the red squiggly for an error will show up in the wrong place in your text editor).

I would be very interested in a solution that could prevent this error at compile time. I can think of a couple ways, but I'm not sure if they are worth the overhead:

1. Add a phantom type variable to the `Position`, `Range` and `QsCompilerDiagnostic` types indicating what their positions are relative to (e.g. statement, specialization, or file). This will make the types more complicated and cumbersome.
2. Prevent direct access to QsLocation.Range, and make an accessor function that automatically adds the position. However, I don't think this would catch forgetting to add the position to an already-existing diagnostic range, such as:
   ```
   { diagnostic with QsCompilerDiagnostic.Range = position + diagnostic.Range }
   ```

Reply:
The other solution to preventing this class of errors would be to make the tests more comprehensive - currently I think the diagnostic tests check that an expected diagnostic is anywhere within a callable or type, but they could be more specific.

Reply:
F# is so proud of its type system. Can we make two incompatible Range types? Like let's say RelativeRange and AbsoluteRange? Like
type AbsoluteRange =
    | AbsoluteRange of Range
type RelativeRange =
    | RelativeRange of Range
And then accept only AbsoluteRange where errors are displayed? Then make + take position and relative range and return absolute range.
1. I have no knowledge of this code, names may be totally wrong.
2. I have no knowledge of F#, the above code may be incorrect.
3. It's up to you to decide if such increase in complexity is worth it.

Reply:
It's a good idea, one problem is that there are two kinds of relative ranges: (1) relative to start of statement and (2) relative to start of callable specialization. Both kinds of relative ranges, as well as an absolute range, can appear in diagnostics generated by the compiler, so I think we would need 3 variants of `QsCompilerDiagnostic`.

One way to deal with this is phantom type variables, so we would have

```fsharp
// 'a not used in the definition - only for creating distinct types.
type 'a Position = ...
type 'a Range = ...
type 'a QsCompilerDiagnostic = ...

QsStatement Position       // Relative to start of statement
QsSpecialization Position  // Relative to start of specialization
File Position              // Relative to start of file, i.e. absolute.

val (+) : QsSpecialization Position -> QsStatement Range -> QsSpecialization Range
val (+) : File Position -> QsSpecialization Range -> File Range
```

This might be worth it - I'll think a bit more about this...

Reply:
So I tried this, and it made the code very complex, because there were a lot of types/functions that needed to contain/return ranges of different types depending on where they are called from. That meant that they needed to add a type variable for the range type. The extra type variables spread throughout the code and it got out of hand pretty quickly.

I think I'm going to abandon this idea for now, but thanks for the suggestion @DmitryVasilevsky!